Run Gramine task on Worker

mcornaton opened the pull request.

Commit:
Run Gramine task on Worker

Commit:
Fix @Service import

Commit:
Fix code smells

Commit:
Add tests

Commit:
Implement `buildPreComputeDockerEnv`

Commit:
Run Gramine task on worker

Commit:
Remove static SMS configuration

Commit:
Fix SPS env var

Commit:
Add bindings to Gramine app runs

james-toussaint commented on src/main/java/com/iexec/worker/compute/app/AppComputeService.java:
```diff
@@ -85,7 +90,7 @@ public AppComputeResponse runCompute(TaskDescription taskDescription,
                 .shouldDisplayLogs(taskDescription.isDeveloperLoggerEnabled())
                 .build();
-        // Enclave should be able to connect to the LAS
+        // Enclave should be able to connect to the LAS/the SPS
```
```suggestion
        // Enclave should be able to connect to the LAS
```
We always need to attach a network to reach a LAS which is only accessible through the local worker network. In production SPS and CAS are accessible over Internet.

james-toussaint commented on src/main/java/com/iexec/worker/compute/post/PostComputeService.java:
```diff
@@ -88,11 +93,14 @@ public boolean runStandardPostCompute(TaskDescription taskDescription) {
         String chainTaskId = taskDescription.getChainTaskId();
+
+        TeeWorkflowConfiguration teeWorkflowConfig =
+                teeWorkflowConfigurationService.getTeeWorkflowConfiguration(taskDescription.getChainTaskId());
```
```suggestion
                teeWorkflowConfigurationService.getTeeWorkflowConfiguration(chainTaskId);
```

mcornaton replied:
Good catch! Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/c405c9cb1327ceb53e2c0c5070905c140d675be5.

james-toussaint commented on src/main/java/com/iexec/worker/tee/TeeAbstractService.java:
```diff
@@ -0,0 +1,26 @@
+import java.util.Collection;
+import java.util.List;
+
+public interface TeeAbstractService {
```
```suggestion
public interface TeeService {
```

mcornaton replied:
Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/adafbaa7e22f8da9f9dce32fcd1113faa76ebfe3.

james-toussaint commented on src/main/java/com/iexec/worker/compute/pre/PreComputeService.java:
```diff
@@ -209,23 +210,27 @@ ReplicateStatusCause teeSessionGenerationErrorToReplicateStatusCause(TeeSessionG
         log.info("Preparing tee input data [chainTaskId:{}]", chainTaskId);
+
+        TeeWorkflowConfiguration teeWorkflowConfig =
+                teeWorkflowConfigurationService.getTeeWorkflowConfiguration(taskDescription.getChainTaskId());
```
```suggestion
                teeWorkflowConfigurationService.getTeeWorkflowConfiguration(chainTaskId);
```

mcornaton replied:
Good catch! Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/c405c9cb1327ceb53e2c0c5070905c140d675be5.

james-toussaint commented on src/main/java/com/iexec/worker/compute/post/PostComputeService.java:
```diff
@@ -88,11 +93,14 @@ public boolean runStandardPostCompute(TaskDescription taskDescription) {
+        TeeWorkflowConfiguration teeWorkflowConfig =
+                teeWorkflowConfigurationService.getTeeWorkflowConfiguration(taskDescription.getChainTaskId());
+
         String postComputeImage = teeWorkflowConfig.getPostComputeImage();
```
NPE?

mcornaton replied:
`TeeWorkflowConfigurationService#getTeeWorkflowConfiguration` returns a non-null value or throws a `RuntimeException` so there's no need to check whether `teeWorkflowConfig` is null.
However, we may have to catch the potential `RuntimeException`. I'll have a look at it.

mcornaton replied:
Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/bf03e5b086f465c0329f7f79c158107a608cc32a. It fixes `TeeWorkflowConfiguration` and `SmsClient` creation.

james-toussaint commented on src/main/java/com/iexec/worker/tee/scone/LasServicesManager.java:
```diff
@@ -0,0 +1,90 @@
+    private final DockerService dockerService;
+
+    private final Map<String, LasService> lasForTask = new HashMap<>();
+    private final Map<String, LasService> startedLasServices = new HashMap<>();
```
In order to avoid spawning LAS clones, when a Scone TEE task is received and after a `TeeWorkflowSharedConfiguration` is received, a new LAS should be started only if no LAS with a same `lasImageUri` exists.
You might use something like `chainTaskIdToLasImageUri` and `lasImageUriToLasService`.

mcornaton replied:
You're right. Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/f9773cd7b5f745a81c346fe4107e8969da8c02b9.

james-toussaint commented on src/main/java/com/iexec/worker/sms/SmsService.java:
```diff
@@ -56,6 +56,13 @@ public Optional<TaskSecrets> fetchTaskSecrets(WorkerpoolAuthorization workerpool
+            return Optional.empty();
+        }
+
+        final SmsClient smsClient = oSmsClient.get();
```
It looks this piece of code is duplicated multiple times. I think we should check during START stage if SMS url is set (if tee task). If unset we can abort the task. If set, the smsClient will never be null (`getOrCreateSmsClient(url)`)

mcornaton replied:
Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/bf03e5b086f465c0329f7f79c158107a608cc32a. It fixes `TeeWorkflowConfiguration` and `SmsClient` creation.

james-toussaint commented on src/main/java/com/iexec/worker/sms/SmsService.java:
```diff
@@ -122,19 +129,23 @@ public void saveSecrets(String chainTaskId,
-                log.error("Failed to get tee workflow configuration from sms", e);
-                teeWorkflowConfiguration = null;
-            }
+    public TeeWorkflowSharedConfiguration getTeeWorkflowConfiguration(String chainTaskId) {
```
If I'm not wrong it looks you have mutiple ways of retrieving `TeeWorkflowConfiguration`: one with `TeeWorkflowConfigurationService` and the other in `SmsService`. `SmsService` might be the useless one

mcornaton replied:
`SmsService` lets you retrieve a `TeeWorkflowSharedConfiguration` while `TeeWorkflowConfigurationService` lets you retrieve a `TeeWorkflowConfiguration` (based on a `TeeWorkflowSharedConfiguration` retrieved through the `SmsService`).

james-toussaint replied:
Ok, so maybe we should rename it to `getTeeWorkflowSharedConfiguration`?

james-toussaint commented on src/main/java/com/iexec/worker/tee/gramine/GramineConfiguration.java:
```diff
@@ -0,0 +1,12 @@
+public class GramineConfiguration {
+    @Getter
+    @Value("${gramine.sps.ssl-certs}")
+    private String sslCertificates;
```
Reminder @mcornaton @jeremyjams

james-toussaint commented on src/main/java/com/iexec/worker/tee/gramine/TeeGramineService.java:
```diff
@@ -0,0 +1,75 @@
+@Service
+@Slf4j
+public class TeeGramineService implements TeeAbstractService {
+    private static final String SPS_URL = "sps";
```
```suggestion
    private static final String SPS_URL_ENVVAR_NAME = "sps";
    .... *_ENVVAR_NAME = ...
```
Or a better suffix

mcornaton replied:
Right, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/40f8c154e00abfb184f92742bdd71d8e84015597.

mcornaton committed:
Remove useless comment

Co-authored-by: Jérémy James Toussaint <33313130+jeremyjams@users.noreply.github.com>

Commit:
Use `chainTaskId` instead of `taskDescription.getChainTaskId()`

Commit:
Rename `TeeAbstractService` into `TeeService`

james-toussaint commented on src/main/java/com/iexec/worker/tee/scone/LasService.java:
```diff
@@ -0,0 +1,95 @@
+        DockerRunRequest dockerRunRequest = DockerRunRequest.builder()
+                .containerName(containerName)
+                .imageUri(imageUri)
+                // application & post-compose enclaves will be
```
```suggestion
                // optional pre-compose, application & post-compose enclaves will be
```

mcornaton replied:
We're talking about `pre-compute` and `post-compute`, not `pre-compose` and `post-compose`, right?
See https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/52e3e325bf706ceb7f2edd3c01d5fc4fadc83f22.

james-toussaint replied:
Yes you are right!

Commit:
Merge branch 'feature/read-gramine-task' of https://github.com/iExecBlockchainComputing/iexec-worker into feature/read-gramine-task

james-toussaint commented on src/main/java/com/iexec/worker/tee/scone/TeeSconeService.java:
```diff
@@ -45,112 +45,81 @@ public class TeeSconeService {
+        String chainTaskId = taskDescription.getChainTaskId();
+        TeeWorkflowConfiguration teeWorkflowConfig =
+                teeWorkflowConfigurationService.getTeeWorkflowConfiguration(chainTaskId);
+        return getDockerEnv(chainTaskId, sconeConfigId, teeWorkflowConfig.getPreComputeHeapSize(), session.getSecretProvisioningUrl());
```
FYI: Me might stop overriding heap size in the future

james-toussaint replied:
Maybe add a TODO in comment?

mcornaton replied:
Sure. It could be removed today, but I don't think it would fit into this PR. Please see the new TODO here: https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/984ae97e3e90ba5b3ab1d0101de261202f514d78.

james-toussaint commented on src/main/java/com/iexec/worker/tee/TeeAbstractService.java:
```diff
@@ -0,0 +1,26 @@
+            TaskDescription taskDescription,
+            @Nonnull TeeSessionGenerationResponse session);
+
+    Collection<String> getBindings();
```
```suggestion
    Collection<String> getAdditionalBindings();
```
?

mcornaton replied:
That's better, right!
https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/2949f847387d47eb5eea70249081712eff129db1.

james-toussaint replied:
Thanks :)

Commit:
Spawn only one `LasService` per `lasImageUri`

Commit:
Rename env vars in Java Code to be more precise about what they are

Commit:
Update comment

Commit:
Add a TODO comment about Scone Heap

Commit:
Rename `TeeService#getBindings` into `TeeService#getAdditionalBindings`

Commit:
Get `SmsClient` & `TeeWorkflowConfig` on task start or abort

james-toussaint commented on src/main/java/com/iexec/worker/executor/TaskManagerService.java:
```diff
@@ -233,6 +268,11 @@ ReplicateActionResponse compute(String chainTaskId) {
                     context, chainTaskId);
         }
 
+        if (taskDescription.isTeeTask() && !lasServicesManager.startLasService(chainTaskId)) {
```
1. This is only true for a Scone task
2. I suggest to move that specific part within the `TeeService`
What do you think?

mcornaton replied:
Good idea, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/75e674171d17f09f8e67bebd6489f7d7bf81b06e.

james-toussaint replied:
Clean :100:

Commit:
Add `TeeService#prepareTeeForTask`

james-toussaint commented on src/main/java/com/iexec/worker/executor/TaskManagerService.java:
```diff
@@ -233,6 +265,14 @@ ReplicateActionResponse compute(String chainTaskId) {
+        if (taskDescription.isTeeTask()) {
+            TeeService teeService = teeServicesManager.getTeeService(taskDescription.getTeeEnclaveProvider());
+            if (!teeService.prepareTeeForTask(chainTaskId)) {
+                return getFailureResponseAndPrintError(LAS_START_FAILED,
```
```suggestion
                return getFailureResponseAndPrintError(TEE_TASK_PREPARATION_FAILED,
```
? (or better if you have)

mcornaton replied:
I think I'd prefer `TEE_PREPARATION_FAILED`. We're not preparing the task here, only the external TEE service(s).

james-toussaint replied:
Better yes :)

Commit:
Refactor `TeeWorkflowConfiguration`

Commit:
Rename `LAS_START_FAILED` into `TEE_PREPARATION_FAILED`

james-toussaint commented on src/main/java/com/iexec/worker/executor/TaskManagerService.java:
```diff
@@ -59,33 +63,39 @@ public class TaskManagerService {
-            SubscriptionService subscriptionService) {
+            SubscriptionService subscriptionService,
+            SmsClientProvider smsClientProvider,
+            TeeWorkflowConfigurationService teeWorkflowConfigurationService) {
```
Constructor is heavy. How about moving some newly introduced logic to `SmsService` and/or `TeeServicesManager`?

mcornaton replied:
It looks better into `TeeService` (which can me changed from an interface to an abstract class). Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/7dd0d140d9802e9dc547b811c3689989c49109f1.

james-toussaint replied:
Yes, we tried to avoid an abstract `TeeService` until here (interfaces are better), but it looks a good compromise for today, thanks :)

james-toussaint replied:
By doing this you can migrate `SgxService` field from `BlaBlaTeeService`s to parent class `TeeService`.

mcornaton replied:
Sure, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/c7338798e2cb37d25e45da4ff4d42ecef658b1d6.

james-toussaint replied:
Nice!
Can you move both
```
    @Override
    public boolean isTeeEnabled() {
        return sgxService.isSgxEnabled();
    }
```
 to `TeeService` while we don't have a third TEE provider?

mcornaton replied:
Sure, see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/d4700ede17b31d1e2609a4821bce19ae1d47eabb.

james-toussaint replied:
Thanks :)

james-toussaint commented on src/main/java/com/iexec/worker/sms/SmsClientConfiguration.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+public class SmsClientConfiguration {
```
```suggestion
public class SmsClientProviderConfiguration {
```

mcornaton replied:
See https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/d8b8dd8a771bd60e607596733ca8e61875863675.

james-toussaint commented on src/main/java/com/iexec/worker/tee/scone/LasServicesManager.java:
```diff
@@ -0,0 +1,87 @@
+    private final SgxService sgxService;
+    private final DockerService dockerService;
+
+    private final Map<String, LasService> chainTaskIdToLasService = new HashMap<>();
```
```suggestion
    //TODO: Purge entry when task is over (completed/failed)
    private final Map<String, LasService> chainTaskIdToLasService = new HashMap<>();
```

Commit:
Move TEE prerequisites check into `TeeService`

Commit:
Rename `SmsClientConfiguration` into `SmsClientProviderConfiguration`

mcornaton committed:
Add TODO about purging map

Co-authored-by: Jérémy James Toussaint <33313130+jeremyjams@users.noreply.github.com>

james-toussaint commented on src/main/java/com/iexec/worker/tee/scone/TeeSconeService.java:
```diff
@@ -45,112 +46,87 @@ public class TeeSconeService {
-            return false;
-        }
-        return true;
+    @Override
```
Pseudocode:
```suggestion
    @Override
    public Optional<ReplicateStatusCause> areTeePrerequisitesMetForTask(String chainTaskId){
        return super.areTeePrerequisitesMetForTask(chainTaskId) && prepareTeeForTask(chainTaskId);
    }
  

    @Override
```
(but also keep `prepareTeeForTask` call in `compute` stage as it is, in case the worker is restarted after `start` stage)

mcornaton replied:
Right, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/499432b5eabece882d2a2cd486879808be66c2df.

james-toussaint replied:
Perfect thanks!

Commit:
Update Scone `areTeePrerequisitesMetForTask` to start LAS

Commit:
Move `sgxService` to `TeeService` & improve code coverage

jbern0rd commented on src/main/java/com/iexec/worker/tee/scone/LasService.java:
```diff
@@ -0,0 +1,95 @@
+        return true;
+    }
+
+    void stop() {
```
Do we keep `isStarted` to `true` when the LAS is stopped ?

mcornaton replied:
I guess we can go back to `false` when stopping the LAS. In theory, it should never be stopped until the end of execution of the worker so no need for it to be toggled. But let's be clean!
https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/ec44ca0df3596f6a639535499cd32cd46b827397.

jbern0rd replied:
hmm, sadly, it's not that clean due to `DockerClientInstance` implementation in iexec-common.
stopAndRemoveContainer implementation can produce weird results because their implementation describes what happens instead of the targeted result.
To give an example, they will return `false` if the provided container has already been removed (for any reason).
This will create a lasStarted value of `true`.
We may need to add a TODO to note the current situation is not good and needs to be reviewed cleanly.

mcornaton replied:
I have tried to resolve the issue. Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/072fecad7d7aa95da56e19e74e4140611b0c83a3.
The idea is to not depend anymore on `DockerClientInstance::stopAndRemoveContainer` and simply check whether the container is still present after the completion of the method.

jbern0rd commented on src/main/java/com/iexec/worker/tee/scone/LasServicesManager.java:
```diff
@@ -0,0 +1,88 @@
+
+        final TeeWorkflowConfiguration config = smsClient.getTeeWorkflowConfiguration();
+        if (config == null) {
+            throw new RuntimeException("Missing tee workflow configuration");
```
This exception does not seem to be caught in the code.
The exception will go up to `start` and `compute` methods in `TaskManagerService` where it will never be caught.
Do we have a potential point where a task lifecycle could be broken ?
Wouldn't it be a safer implementation with a log message and a returned `false` literal ?

mcornaton replied:
You're absolutely right. Please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/38390766d3f2bd11701ff8ae2bb560f26f78ccbd.

jbern0rd commented on src/test/java/com/iexec/worker/compute/post/PostComputeServiceTests.java:
```diff
@@ -27,12 +27,13 @@
 import com.iexec.sms.api.TeeSessionGenerationResponse;
 import com.iexec.worker.compute.ComputeExitCauseService;
-import com.iexec.worker.compute.TeeWorkflowConfiguration;
+import com.iexec.sms.api.TeeWorkflowConfiguration;
```
Little issue in import order here

mcornaton replied:
Right, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/2f68229c991043391f6f042d5760d892616a5b8e.

jbern0rd commented on src/test/java/com/iexec/worker/compute/pre/PreComputeServiceTests.java:
```diff
@@ -26,18 +26,20 @@
 import com.iexec.sms.api.TeeSessionGenerationResponse;
 import com.iexec.worker.compute.ComputeExitCauseService;
-import com.iexec.worker.compute.TeeWorkflowConfiguration;
+import com.iexec.sms.api.TeeWorkflowConfiguration;
```
Little issue in imports order here

mcornaton replied:
Right, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/2f68229c991043391f6f042d5760d892616a5b8e.

Commit:
Reorder imports

Commit:
Toggle `isStarted` when stopping `LasService`

Commit:
Remove cert mapping

Commit:
Merge branch 'feature/read-gramine-task' of github.com:iExecBlockchainComputing/iexec-worker into feature/read-gramine-task

Commit:
Move `isTeeEnabled` to `TeeService`

Commit:
Return false instead of throwing exception is `startLasService`

+ add tests

Commit:
Fix LasService double start

Commit:
Update `isStarted` value in `LasService`

jbern0rd commented on src/main/java/com/iexec/worker/compute/pre/PreComputeService.java:
```diff
@@ -80,8 +81,6 @@ public PreComputeService(
      * is started to handle them.
      *
-     * @param taskDescription
-     * @param workerpoolAuth
```
Why were those parameters removed from the documentation ?

mcornaton replied:
Not sure. Here they are restored: https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/5a6da992ecb571de4b165cbedff0dafa26a66039.

james-toussaint commented on src/main/java/com/iexec/worker/tee/TeeWorkflowConfigurationService.java:
```diff
@@ -0,0 +1,73 @@
+        final String preComputeImage = config.getPreComputeImage();
+        final String postComputeImage = config.getPostComputeImage();
+
+        if (!dockerService.getClient(preComputeImage)
```
I don't think force pull is required here.
If image is not present, pull it, else keep going

mcornaton replied:
I disagree. In case an image is rebuilt and uses the same tag as before - bad practice, but we sometimes have this issue -, then we'd like to pull.
BTW, I'm pretty sure Docker will only pull missing layers, not the complete image. So if we already have all the layers, it will simply be a fast check, right?

james-toussaint replied:
FYI:
"Docker will only pull missing layers" true
"it will simply be a fast check" it depends how the image is built (example, if an apt update/upgrade is made in first layers, next layers will be all different)

mcornaton replied:
Right, please see https://github.com/iExecBlockchainComputing/iexec-worker/pull/464/commits/50f398c3b157f77bd60123395e673a5b71814d56.

Commit:
Restore Java doc parameters

Commit:
Do not force-pull pre/post-compute images

Commit:
Update `LasService::stopAndRemoveContainer`

Setting a coherent return value, not depending on `DockerClientInstance.stopAndRemoveContainer`

Commit:
Add //TODO to `stopAndRemoveContainer` methods
